File: src/main/osu/beatmap/sections/BeatmapEvents.java
package main.osu.beatmap.sections;

import java.util.ArrayList;

import main.osu.RGBColor;
import main.osu.game.BreakPeriod;

/**
 * Contains beatmap events.
 */
public class BeatmapEvents {
    /**
     * The file name of this beatmap's background.
     */
    public String backgroundFilename;

    /**
     * The breaks this beatmap has.
     */
    public ArrayList<BreakPeriod> breaks = new ArrayList<>();

    /**
     * The background color of this beatmap.
     */
    public RGBColor backgroundColor;

    public BeatmapEvents() {}

    /**
     * Copy constructor.
     *
     * @param source The source to copy from.
     */
    private BeatmapEvents(BeatmapEvents source) {
        backgroundFilename = source.backgroundFilename;

        for (BreakPeriod breakPeriod : source.breaks) {
            breaks.add(new BreakPeriod(breakPeriod.getStart(), breakPeriod.getEndTime()));
        }

        backgroundColor = source.backgroundColor != null ? new RGBColor(backgroundColor) : null;
    }

    /**
     * Deep clones this instance.
     *
     * @return The deep cloned instance.
     */
    public BeatmapEvents deepClone() {
        return new BeatmapEvents(this);
    }
}

File: src/com/rian/osu/beatmap/sections/BeatmapEvents.kt
package com.rian.osu.beatmap.sections

import ru.nsu.ccfit.zuev.osu.RGBColor
import com.rian.osu.beatmap.timings.BreakPeriod

/**
 * Contains beatmap events.
 */
class BeatmapEvents : Cloneable {
    /**
     * The file name of this beatmap's background.
     */
    @JvmField
    var backgroundFilename: String? = null

    /**
     * The file name of this beatmap's background video.
     */
    @JvmField
    var videoFilename: String? = null

    /**
     * The beatmap's background video start time in milliseconds.
     */
    @JvmField
    var videoStartTime = 0

    /**
     * The breaks this beatmap has.
     */
    @JvmField
    var breaks = mutableListOf<BreakPeriod>()

    /**
     * The background color of this beatmap.
     */
    @JvmField
    var backgroundColor: RGBColor? = null

    public override fun clone() =
        (super.clone() as BeatmapEvents).apply {
            this@BeatmapEvents.breaks.forEach { breaks.add(it.copy()) }
            backgroundColor =
                if (this@BeatmapEvents.backgroundColor != null) RGBColor(this@BeatmapEvents.backgroundColor)
                else null
        }
}
